### PR TITLE
Use a composite key of all bundle ids and version to check the cache

### PR DESCRIPTION
**Pull request by laeubi**
Currently the ContentDescriptionManager relies on a rather vague "timestamp" to indicate that the set of bundles has changed, this leads to some hard to explain instabilities in tests.

This now uses a composite key of all bundle ids and their version that has two benefits:

1) it is deterministic what properties of a bundle set is compared
2) in the test we can assert that the state is stable and if not give a comparison list that can be used to determine what actually has changed

### CODE REVIEW

**Comment by laeubi** [comment]
okay it seems they stored data is limited to 2048 bytes .... so actually some kind of Hash would be required, for testing purpose I raised the limit to see if we can find if/what is failing the state compare here...

**Comment by laeubi** [comment]
Alright, now large properties test fails but 2ndsession succeeds, I'll one more adjust this and then it hopefully is gone.

File: resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ContentDescriptionManager.java [hunk 1]
```diff
@@ -65,6 +73,10 @@
  * @see IFile#getContentDescription()
  */
 public class ContentDescriptionManager implements IManager, IRegistryChangeListener, IContentTypeManager.IContentTypeChangeListener, ILifecycleListener {
+	private static final String EMPTY = ""; //$NON-NLS-1$
+	private static final String DELIMITER = System.lineSeparator();
+	private static final String SPACE = " "; //$NON-NLS-1$
```
Comment (HeikoKlare) [inline resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ContentDescriptionManager.java:78]: One nitpicky design-related comment: these constants are a bit confusing, as they do not belong to the `ContentDescriptionManager` but to some kind of `PlatformStateDescription` that is handled by `getCurrentPlatformState()`, `getCachedPlatformState()` and the property setter in `shutdown()`. In order to separate concerns and clarify where these constants belong to, you could encapsulate this logic in some (inner, static) `PlatformStateDescription` class.

File: resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ContentDescriptionManager.java [hunk 2]
```diff
@@ -480,6 +482,37 @@ public synchronized void invalidateCache(boolean flush, IProject project) {
...
+			return HexFormat.of().formatHex(digest.digest(ID.getBytes(StandardCharsets.UTF_8)));
+		} catch (NoSuchAlgorithmException e) {
+			return Integer.toHexString(ID.hashCode());
+		}
+	}
+
+	/**
+	 * @return the cached platform state, only public for testing purpose!
+	 */
+	public String getCachedPlatformState() {
```
Comment (HeikoKlare) [inline resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ContentDescriptionManager.java:508]: Currently there is not test requiring this, so probably this method can be private?

**Review by HeikoKlare — COMMENTED** [review]
The changes look sound, as the state description should now contain exactly what is required to identify a cache-relevant state change. Thanks for working on this!

I only have two proposals in terms of design, but I am also fine with the changes as they are.

File: resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ContentDescriptionManager.java (hunk 1 above)
Comment (laeubi) [reply resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ContentDescriptionManager.java:78]: Well actually this just to not make the code scattered with all these `//$NON-NLS-1$` stuff... I can't get why one ever would want to "translate" an empty String or a single space but I'm sure there is a reason I just don't understand :-P

File: resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ContentDescriptionManager.java (hunk 2 above)
Comment (laeubi) [reply resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ContentDescriptionManager.java:508]: yes I missed to revert that...

**Review by HeikoKlare — APPROVED** [review]
Looks good, thanks!
Let's see if this finally solves the issue :-)

**Pull request merged**